api/src/Repository/Model/BaseModel.php

```diff
@@ -0,0 +1,277 @@
+        $models = $this->isCollection($models)?$models->all(true):[$models];
+        $this->deleteStack = array_merge($this->deleteStack, array_filter($models));
+    }
+
+
+    /**
+     * Save the model and all of its relationships.
+     *
+     * @return bool
+     */
+    public function push()
+    {
+        if (!$this->save()) {
+            return false;
+        }
+
+        foreach ($this->deleteStack as $modelToDelete) {
+            if (!$modelToDelete->delete()) {
+                return false;
+            }
+            $this->deleteStack = [];
+        }
+
+        // To sync all of the relationships to the database, we will simply spin through
+        // the relationships and save each model via this "push" method, which allows
+        // us to recurse into all of these nested relations for the model instance.
+        foreach ($this->relations as $key => $models) {
+            /** @var Collection|array $models */
+            $models = $this->isCollection($models)? $models->all(true) : [$models];
+            $relation = static::$relationsCache[$this->getRelationCacheKey($key)];
+            foreach (array_filter($models) as $model) {
+                /** @var BaseModel $model */
+                $model->preserveKeys($relation);
+                if (!$model->push()) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Checks relation against value
+     * @param $method
+     * @param $value
+     * @return bool
+     * @throws SetRelationException
+     */
+    protected function isValueCompatibleWithRelation($method, $value)
+    {
+        if (is_null($value)) {
+            return true;
+        }
+
+        $relation = static::$relationsCache[$this->getRelationCacheKey($method)];
+
+        if ($relation instanceof HasOne || $relation instanceof BelongsTo) {
+            if ($this->isCollection($value)) {
+                throw new SetRelationException('Can not set collection, model expected');
+            }
+        } else {
+            if ($this->isModel($value)) {
+                throw new SetRelationException('Can not set model, collection expected');
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @param $value
+     * @return bool
+     */
+    protected function isModel($value)
+    {
+        return $value instanceof BaseModel;
+    }
+
+    /**
+     * @param $value
+     * @return bool
+     */
+    protected function isCollection($value)
+    {
+        return $value instanceof Collection;
+    }
+
+    /**
+     * @param Relation $relation
+     */
+    protected function preserveKeys(Relation $relation)
+    {
+        if ($relation instanceof HasOneOrMany) {
+            $fk = str_replace($this->getTable().'.', '', $relation->getForeignKey());
+            $this->{$fk} = $relation->getParentKey();
+        }
+
+//        if ($relation instanceof BelongsTo){
```

Comment:
yeah, I will add more comments on this.
It's TODO mostly, as belongsTo has a different interface and I'm not sure if we even want to implement this